src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -248,23 +259,43 @@ public void Reset()
             _abortedCts = null;
         }
 
-        protected void ResetComponents()
+        protected void ResetComponents(bool immediateReuse)
         {
-            var frameHeaders = Interlocked.Exchange(ref _frameHeaders, null);
-            if (frameHeaders != null)
+            if (immediateReuse && HeaderFactory != null)
```

Comment:
I think the Reset method should go via the factory and the factory should return the "new" component; as it is still making an assumption that when there is a factory it is pooling here; whereas it might be doing something else like tracking, monitoring, counting etc.